File: apps/api-python/tests/quality_db.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path
from urllib.parse import urlparse, urlunparse

PROTECTED_DB_NAMES = frozenset({"asa", "postgres", "template0", "template1"})
API_PYTHON_ROOT = Path(__file__).resolve().parents[1]


def _fail_env(message: str) -> None:
    raise RuntimeError(
        f"Quality DB environment failure: {message}. "
        "Ensure Postgres is running and reachable "
        "(e.g. postgresql://asa:asa@localhost:5432/postgres)."
    )


def quality_database_name() -> str:
    # Separate from the Node quality DB so Prisma/Alembic do not race on one catalog.
    return os.environ.get("QUALITY_DATABASE_NAME", "asa_quality_test_python")


def resolve_admin_database_url() -> str:
    if os.environ.get("QUALITY_ADMIN_DATABASE_URL"):
        return os.environ["QUALITY_ADMIN_DATABASE_URL"]
    from_env = os.environ.get("DATABASE_URL")
    if from_env:
        parsed = urlparse(from_env)
        return urlunparse(parsed._replace(path="/postgres"))
    return "postgresql://asa:asa@localhost:5432/postgres"


def resolve_quality_database_url() -> str:
    if os.environ.get("QUALITY_DATABASE_URL"):
        return os.environ["QUALITY_DATABASE_URL"]
    admin = resolve_admin_database_url()
    parsed = urlparse(admin)
    return urlunparse(parsed._replace(path=f"/{quality_database_name()}"))


def _psql(admin_url: str, *args: str) -> str:
    try:
        completed = subprocess.run(
            ["psql", admin_url, "-v", "ON_ERROR_STOP=1", *args],
            check=True,
            capture_output=True,
            text=True,
        )
        return completed.stdout
    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
        detail = getattr(exc, "stderr", None) or str(exc)
        _fail_env(f"psql failed: {detail}")
        raise  # pragma: no cover


def _db_exists_sql(name: str) -> str:
    escaped = name.replace("'", "''")
    return f"SELECT 1 FROM pg_database WHERE datname = '{escaped}'"


def ensure_quality_database() -> str:
    """Create asa_quality_test if needed and run Alembic migrations. Never touches `asa`."""
    name = quality_database_name()
    if name in PROTECTED_DB_NAMES:
        _fail_env(f'refusing to use protected database name "{name}"')

    admin_url = resolve_admin_database_url()
    existing = _psql(admin_url, "-tAc", _db_exists_sql(name)).strip()
    if existing != "1":
        try:
            _psql(admin_url, "-c", f'CREATE DATABASE "{name}"')
        except RuntimeError:
            again = _psql(admin_url, "-tAc", _db_exists_sql(name)).strip()
            if again != "1":
                raise

    quality_url = resolve_quality_database_url()
    if quality_url.rstrip("/").endswith("/asa") or "/asa?" in quality_url:
        _fail_env("refusing to point quality tests at the asa database")

    os.environ["DATABASE_URL"] = quality_url

    try:
        subprocess.run(
            [
                str(API_PYTHON_ROOT / ".venv" / "bin" / "alembic"),
                "upgrade",
                "head",
            ],
            cwd=API_PYTHON_ROOT,
            check=True,
            capture_output=True,
            text=True,
            env={**os.environ, "DATABASE_URL": quality_url},
        )
    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
        detail = getattr(exc, "stderr", None) or str(exc)
        _fail_env(f"alembic upgrade failed: {detail}")

    return quality_url

File: templates/api-node/src/http/quality-db.ts
import { execFileSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const PROTECTED_DB_NAMES = new Set(["asa", "postgres", "template0", "template1"]);

function failEnv(message: string): never {
  throw new Error(
    `Quality DB environment failure: ${message}. Ensure Postgres is running and reachable (e.g. postgresql://asa:asa@localhost:5432/postgres).`,
  );
}

function readEnv(key: string): string | undefined {
  return process.env[key];
}

function writeEnv(key: string, value: string): void {
  process.env[key] = value;
}

export function qualityDatabaseName(): string {
  // Separate from the Python quality DB so Prisma/Alembic do not race on one catalog.
  return readEnv("QUALITY_DATABASE_NAME") ?? "asa_quality_test_node";
}

export function resolveAdminDatabaseUrl(): string {
  const adminOverride = readEnv("QUALITY_ADMIN_DATABASE_URL");
  if (adminOverride) {
    return adminOverride;
  }
  const fromEnv = readEnv("DATABASE_URL");
  if (fromEnv) {
    try {
      const url = new URL(fromEnv);
      url.pathname = "/postgres";
      return url.toString();
    } catch {
      // fall through
    }
  }
  return "postgresql://asa:asa@localhost:5432/postgres";
}

export function resolveQualityDatabaseUrl(): string {
  const override = readEnv("QUALITY_DATABASE_URL");
  if (override) {
    return override;
  }
  const admin = resolveAdminDatabaseUrl();
  try {
    const url = new URL(admin);
    url.pathname = `/${qualityDatabaseName()}`;
    return url.toString();
  } catch {
    return `postgresql://asa:asa@localhost:5432/${qualityDatabaseName()}`;
  }
}

function psql(adminUrl: string, args: string[]): string {
  try {
    return execFileSync("psql", [adminUrl, "-v", "ON_ERROR_STOP=1", ...args], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
    });
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    failEnv(`psql failed: ${detail}`);
  }
}

/**
 * Create the Node quality DB if missing and run Prisma migrate deploy.
 * Never drops or truncates the `asa` development database.
 */
export async function ensureQualityDatabase(): Promise<string> {
  const name = qualityDatabaseName();
  if (PROTECTED_DB_NAMES.has(name)) {
    failEnv(`refusing to use protected database name "${name}"`);
  }

  const adminUrl = resolveAdminDatabaseUrl();
  const existing = psql(adminUrl, [
    "-tAc",
    `SELECT 1 FROM pg_database WHERE datname = '${name.replace(/'/g, "''")}'`,
  ]).trim();

  if (existing !== "1") {
    try {
      psql(adminUrl, ["-c", `CREATE DATABASE "${name}"`]);
    } catch (error) {
      const again = psql(adminUrl, [
        "-tAc",
        `SELECT 1 FROM pg_database WHERE datname = '${name.replace(/'/g, "''")}'`,
      ]).trim();
      if (again !== "1") {
        throw error;
      }
    }
  }

  const qualityUrl = resolveQualityDatabaseUrl();
  if (qualityUrl.includes("/asa?") || qualityUrl.endsWith("/asa")) {
    failEnv("refusing to point quality tests at the asa database");
  }

  writeEnv("DATABASE_URL", qualityUrl);

  const apiRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
  try {
    execFileSync(path.join(apiRoot, "node_modules/prisma/build/index.js"), ["migrate", "deploy"], {
      cwd: apiRoot,
      encoding: "utf8",
      env: { ...process.env, DATABASE_URL: qualityUrl },
      stdio: ["ignore", "pipe", "pipe"],
    });
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    failEnv(`prisma migrate deploy failed: ${detail}`);
  }

  return qualityUrl;
}
